### ISSUE TITLE
How to use the model for inference?

### ISSUE THREAD

**Issue by asif-nuaa** [lartpang/ZoomNeXt#7]
Please elaborate more in Readme about how to use the code for inference or testing. which folder do we need to place the pretrained weights .pth files. procedure for finetuning on new dataset if possible.

**Comment by asif-nuaa** [comment]
I am attempting to use pre-trained weights for inference on a set of images, but I am encountering issues with the command-line arguments and configuration. My goal is to use the model to produce **resulting images** of concealed objects on COD10k images or from any other dataset images.

**Command Used:**
```
python main_for_image.py --config configs/icod_train.py --model-name PvtV2B5_ZoomNeXt --evaluate --load-from pvtv2-b5-zoomnext.pth
```

**Configuration Files:**

'icod_train.py:'
```
data=dict(
    shape=dict(h=384, w=384),
    names=["cod10k_tr"],
)

test = dict(
    batch_size=__BATCHSIZE,
    num_workers=2,
    clip_range=None,
    data=dict(
        shape=dict(h=384, w=384),
        names=["cod10k_te"],
    ),
)
```

**dataset.yaml:**

```
# ICOD Datasets
cod10k_tr:
  {
    root: "data/COD10K-v3/Train/",
    image: { path: "Image", suffix: ".jpg" },
    mask: { path: "GT_Object", suffix: ".png" },
  }

cod10k_te:
  {
    root: "data/COD10K-v3/Test/",
    image: { path: "Image", suffix: ".jpg" },
    mask: { path: "GT_Object", suffix: ".png" },
  }
```

**Log Snippet**
```
(OtitisNet2) PS D:\asif\Projects\OtitisNet2\ZoomNeXt-main> python main_for_image.py --config configs/icod_train.py --model-name PvtV2B5_ZoomNeXt --e
valuate --load-from pvtv2-b5-zoomnext.pth
INFO:albumentations.check_version:A new version of Albumentations is available: 1.4.13 (you have 1.4.12). Upgrade using: pip install -U albumentatio
ns. To disable automatic update checks, set the environment variable NO_ALBUMENTATIONS_UPDATE to 1.
[main_for_image.py] base_seed = 112358
config = 'configs/icod_train.py'
data_cfg = './dataset.yaml'
dataset_infos = dict(
    cod10k_te=dict(
        image=dict(path='Image', suffix='.jpg'),
        mask=dict(path='GT_Object', suffix='.png'),
        root='data/COD10K-v3/Test/'),
    cod10k_tr=dict(
        image=dict(path='Image', suffix='.jpg'),
        mask=dict(path='GT_Object', suffix='.png'),
        root='data/COD10K-v3/Train/'))
deterministic = True
device = 'cuda:0'
evaluate = True
exp_name = 'PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP'
has_test = True
info = None
load_from = 'pvtv2-b5-zoomnext.pth'
log_interval = 20
metric_names = [
    'sm',
    'wfm',
    'mae',
    'em',
    'fmeasure',
]
model_name = 'PvtV2B5_ZoomNeXt'
output_dir = 'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs'
path = dict(
    cfg_copy=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\co
nfig.py',
    excel=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\re
sults.xlsx',
    final_full_net=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\pt
h\\checkpoint_final.pth',
    final_state_net=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\pt
h\\state_final.pth',
    log=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\lo
g_2024-08-07.txt',
    output_dir='D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs',
    pth=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\pt
h',
    pth_log=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4',  
    save=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\pr
e',
    tb=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\tb
',
    trainer_copy=
    'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main\\outputs\\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\\exp_4\\tr
ainer.txt'
)
pretrained = False
proj_root = 'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main'
save_results = False
test = dict(
    batch_size=4,
    clip_range=None,
    data=dict(names=[
        'cod10k_te',
    ], shape=dict(h=384, w=384)),
    num_workers=2)
train = dict(
    batch_size=4,
    bn=dict(freeze_affine=True, freeze_encoder=False, freeze_status=True),
    data=dict(names=[
        'cod10k_tr',
    ], shape=dict(h=384, w=384)),
    epoch_based=True,
    grad_acc_step=1,
    lr=0.0001,
    num_epochs=150,
    num_iters=None,
    num_workers=2,
    optimizer=dict(
        cfg=dict(diff_factor=0.1, weight_decay=0),
        group_mode='finetune',
        mode='adam',
        set_to_none=False),
    sche_usebatch=True,
    scheduler=dict(
        cfg=dict(gamma=0.1, milestones=101000),
        mode='step',
        warmup=dict(initial_coef=0.01, mode='linear', num_iters=0)),
    use_amp=True)
use_custom_worker_init = True

[pt_utils.py] We will use a fixed seed 112358
[pt_utils.py] We will not use `torch.backends.cudnn.benchmark`
[main_for_image.py] class PvtV2B5_ZoomNeXt(PvtV2B2_ZoomNeXt):
    def set_backbone(self, pretrained: bool, use_checkpoint: bool):
        self.encoder = pvt_v2_eff_b5(pretrained=pretrained, use_checkpoint=use_checkpoint)

[main_for_image.py] Number of Parameters: 84773821
Saving weight 'D:\asif\Projects\OtitisNet2\ZoomNeXt-main\outputs\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\exp_4\
pth\state_final.pth'
Saved weight 'D:\asif\Projects\OtitisNet2\ZoomNeXt-main\outputs\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP\exp_4\p
th\state_final.pth' (only contain the net's weight)
[main_for_image.py] Testing with testset: cod10k_te: 4000
[EVAL]:   0%|                                         | 0/1000 [00:00<?, ?it/s]INFO:albumentations.check_version:A new version of Albumentations is 
available: 1.4.13 (you have 1.4.12). Upgrade using: pip install -U albumentations. To disable automatic update checks, set the environment variable 
NO_ALBUMENTATIONS_UPDATE to 1.
INFO:albumentations.check_version:A new version of Albumentations is available: 1.4.13 (you have 1.4.12). Upgrade using: pip install -U albumentatio
ns. To disable automatic update checks, set the environment variable NO_ALBUMENTATIONS_UPDATE to 1.
[EVAL]: 100%|██████████████████████████████| 1000/1000 [10:39<00:00,  1.56it/s]
[main_for_image.py] (cod10k_te): {
    root: data/COD10K-v3/Test/
    image: {'path': 'Image', 'suffix': '.jpg'}
    mask: {'path': 'GT_Object', 'suffix': '.png'}
}
sm: 0.876, wfm: 0.419, mae: 0.082, avgem: 0.896, adpem: 0.862, maxem: 0.905, avgfmeasure: 0.430, adpfmeasure: 0.412, maxfmeasure: 0.440
[main_for_image.py] End training...
```

**Questions:**
1-    Is there a way to run the model in inference mode only for qualitative comparison purpose?
2-    How can I modify the configuration to directly process a single image or a set of images, to generate resulting images?

**Comment by lartpang** [comment]
Hi, @asif-nuaa 

1. The evaluation of the method is beyond the scopy of this repository. It is just for training the model and a simple testing. So, If you want to evaluate the model performence, please follow the [description](https://github.com/lartpang/ZoomNeXt/tree/main?tab=readme-ov-file#evaluation) and use the specific toolkit [PySODEvalToolkit](https://github.com/lartpang/PySODEvalToolkit) to do it.
2. For generating the predicion results, just add the following item for image and video tasks and the predictions will be saved into the output dir:
   - https://github.com/lartpang/ZoomNeXt/blob/c3fc51c8c7e8de77e791362485a9a2dea20c8ea0/main_for_image.py#L347
   - https://github.com/lartpang/ZoomNeXt/blob/c3fc51c8c7e8de77e791362485a9a2dea20c8ea0/main_for_video.py#L469

**Comment by asif-nuaa** [comment]
To generate the predicted mask, do I need the PySODEvalToolkit?

Also, I noticed that in the ZoomNeXt/main_for_image.py file, the **--save-results** flag is already set to True on line 347:

```
parser.add_argument("--save-results", action="store_true")
```

However, in the configuration file **config.py** in output folder, the **save_results** parameter is **False**:

```
config.py
---------
base_seed = 112358
config = 'configs/icod_train.py'
data_cfg = './dataset.yaml'
dataset_infos = dict(
    cod10k_te=dict(
        image=dict(path='Image', suffix='.jpg'),
        mask=dict(path='GT_Object', suffix='.png'),
        root='data/COD10K-v3/Test/'),
    cod10k_tr=dict(
        image=dict(path='Image', suffix='.jpg'),
        mask=dict(path='GT_Object', suffix='.png'),
        root='data/COD10K-v3/Train/'))
   ...
   ...
   ...
pretrained = False
proj_root = 'D:\\asif\\Projects\\OtitisNet2\\ZoomNeXt-main'
save_results = False
```

The output directory (...\ZoomNeXt-main\outputs\PvtV2B5_ZoomNeXt_BS4_LR0.0001_E150_H384_W384_OPMadam_OPGMfinetune_SCstep_AMP) contains five subfolders (exp_0, exp_1, exp_2, exp_3, exp_4), each with pre and pth folders, and the files config.py, log_2024-08-07.txt, and trainer.txt. config.py shown above. 

Can you clarify if these settings are conflicting or if there is a need to adjust the save_results parameter in config.py and how?

**Comment by lartpang** [comment]
Thanks! @asif-nuaa 

I will check these code to find the imperceptible 🐛 ...

**Comment by lartpang** [comment]
@asif-nuaa Oh, there may be a little ambiguity here. 

Actually, the setting `action="store_true"` defaults to False, so you'll need to enable this option `--save-results` in your command to use this function.

**Comment by asif-nuaa** [comment]
Thanks for your help! The model is now generating predicted masks as I desired. I have one more question: How can I use this model to generate a prediction mask for a single image without evaluating metrics? Specifically, I want to use the model for inference on just one image and obtain its prediction mask.

I noticed that the **test_wrapper** instance of the **Evaluator** class processes a batch of images from a data loader (main_for_image.py). However, I'm only interested in generating a prediction mask for a single image. How should I modify the code or approach to achieve this?

**Comment by lartpang** [comment]
Hi, @asif-nuaa 

In this repository, we only implement the function of batch-wise training and testing for image and video.
For predicting a single image, you need to modify the main script.
The core function for this need has been written in the function `test` containing `ImageTestDataset` for loading data and `Evaluator` for evaluating predictions.
- Convert the batch-wise data loading by `ImageTestDataset` to a simple method for loading a single image, referring to the `__getitem__` method of `ImageTestDataset`.
- Convert the batch-wise evaluation in `eval` of `Evaluator` to a loop based on a single sample, and remove the useless code about `GroupedMetricRecorder` in the prediciton process.
